### ISSUE TITLE
Command#join not loading full STDOUT output

### ISSUE THREAD

**Issue by kevindew** [xeger/backticks#17]
Hi there,

We've found ourselves dealing with an intermittent issue in this gem where a command that produces more than 1,024 of output seems to fall pray to a race condition when Command#join is called.

The code in question is this: https://github.com/xeger/backticks/blob/20820e3518e1563a84102d2962782824b452cf7f/lib/backticks/command.rb#L98-L105

And our theory is that when `capture` runs it collects the first 1,024 characters of stdout, then when `Process.waitpid(@pid, Process::WNOHANG)` runs it finds the process completed so the remaining characters of stdout are never captured. I'm presuming this is intermittent as sometimes when this runs the `Process.waitpid(@pid, Process::WNOHANG)` will be still running and cause the loop to repeat. I was able to replicate this problem consistently by placing a `sleep` between `capture` and `waitpid` to create an artificial delay.

The exact circumstances where we're seeing this is through a different gem you've authored, https://github.com/xeger/docker-compose, where we're running `Docker::Compose.new.ps` and receiving a truncated output that causes an exception at: https://github.com/xeger/docker-compose/blob/c1428491b686a33cbafdc26e609e6a863e2ec6fa/lib/docker/compose/session.rb#L285

with output of:

```
13:16:58  Docker::Compose::Error: 'docker ps' failed with status 'pid 15874 exit 0': Cannot parse output: '(a7a18ebdc64e1dd0a40f68bd62aa4b1126c68fb9e945a6857ede159e377da519) (jwilder/nginx-proxy:latest) (18.2kB (virtual 158MB)) (Up 27 seconds) (lishing-e2e-tests_test-against_2_nginx-proxy_1) (com.docker.compose.project=lishing-e2e-tests_test-against_2,org.opencontainers.image.created=2021-04-12T00:31:10.992Z,org.opencontainers.image.licenses=MIT,org.opencontainers.image.url=https://github.com/nginx-proxy/nginx-proxy,org.opencontainers.image.version=latest,com.docker.compose.config-hash=5517769695a167132a40ede6634a33e23c406504e4697fc1ff2f8924a11aba09,com.docker.compose.oneoff=False,com.docker.compose.version=1.22.0,org.opencontainers.image.source=https://github.com/nginx-proxy/nginx-proxy,org.opencontainers.image.title=nginx-proxy,maintainer=Jason Wilder mail@jasonwilder.com,org.opencontainers.image.description=Automated nginx proxy for Docker containers using docker-gen,org.opencontainers.image.revision=6613e272eb41c926605bc0411b7af9f70c2ebeee,com.docker.compose.container-number=1,com.docker.compose.service=nginx-'
```

**Comment by xeger** [comment]
Oh boy; this one doesn't sound fun to debug. Thank you for the detailed bug report and sorry it took me so long to get back to you. I will have a look over the weekend .. it sounds to me like pipe buffering is an issue (the 1024-byte thing is a very good clue) and pipe buffering is notoriously finnicky between various unixes..

TL;DR I may not be able to provide a systemic fix.

**Comment by xeger** [comment]
Thank you @kevindew for the bug report; I finally have some downtime to maintain old OSS projects and this one was easy to nail down with your superlative reporting.

I'm still glad you stopped using `docker-compose`; it's a relic from the days when compose didn't have a proper API, and it's over-designed for trivial use cases (or really, most use cases).
